### DISCUSSION TITLE
Outline and Keycloak

### DISCUSSION THREAD

**Discussion by AndreaPersico** [outline/outline#3996] [Self Hosting]
Hello, 
I'm facing an error with Keycloak auth method on a self-hosted environment while logging-in: 

![image](https://user-images.githubusercontent.com/49158551/185783949-0a68e455-70c6-4432-b538-2196b2d7b43d.png)

`{"error":"An email field was not returned in the profile parameter, but is required.","stack":"UnauthorizedError: An email field was not returned in the profile parameter, but is required.\n    at AuthenticationError (/opt/outline/build/server/errors.js:35:34)\n    at OAuth2Strategy._verify (/opt/outline/build/server/routes/auth/providers/oidc.js:72:47)\n    at /opt/outline/node_modules/passport-oauth2/lib/strategy.js:196:24\n    at OAuth2Strategy._passportOauth.Strategy.userProfile (/opt/outline/build/server/routes/auth/providers/oidc.js:44:12)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)","level":"error","message":"Error during authentication"}`

The relevant part of the outline's configuration:

      OIDC_CLIENT_ID: outline
      OIDC_CLIENT_SECRET: <my_secret>
      OIDC_AUTH_URI: <my_uri>/auth/realms/master/protocol/openid-connect/auth
      OIDC_TOKEN_URI: <my_uri>/auth/realms/master/protocol/openid-connect/token
      OIDC_USERINFO_URI: <my_uri>/auth/realms/master/protocol/openid-connect/userinfo
      OIDC_USERNAME_CLAIM: email
      OIDC_DISPLAY_NAME: OpenID
      OIDC_SCOPES: email 

Any suggestion?

**Comment by julian-forster** [comment]
I had the same issue when connecting a self hosted gitlab instance yesterday. Although I set all permissions for the app, i still got the error. When I switched to azure, i got it working.
The only thing you can try, is adding openid to the scopes. Maybe this helps.

**Comment by k1ck3r** [comment]
Had the same issue with configuring Gitea, and removing the oidc scopes worked for me

Working with:
```    - OIDC_AUTH_URI=https://redacted/login/oauth/authorize
      - OIDC_TOKEN_URI=https://redacted/login/oauth/access_token
      - OIDC_USERINFO_URI=https://redacted/login/oauth/userinfo
      - OIDC_CLIENT_ID=XXXX
      - OIDC_CLIENT_SECRET=XXXX
      - OIDC_USERNAME_CLAIM=preferred_username
      - OIDC_DISPLAY_NAME=Gitea
```

hope this help you further with gitlab/keycloak

**Reply by djmaze** [reply]
After [this change](https://github.com/outline/outline/commit/7c47ab560e31f9d16587e30d9a46a17e73a7b32e) oidc login via Gitea does not work anymore for me. Now getting the error `A name field was not returned in the profile parameter, but is required.` all the time.

Anyone else experiencing this?

**Reply by djmaze** [reply]
Oh well, turns out I needed to fill out the (optional) "full name" field in my Gitea profile to make this work. Not sure this correct behaviour from Gitea and/or Outline..

**Comment by AndreaPersico** [comment]
Thanks everyone!
The issue was simpler: Keycloak administrative user did NOT have an email in the profile data 😓 and I was expecting to log in with a different user with the populated data. 
Removing the Clients active Session and using the right user (with a valid email field in the profile) for the authentication was the solution.
Silly me.

**Reply by Wlmrsi** [reply]
Hi @AndreaPersico 

I tried to use Keycloak (v19.0.2) as my authentication server but always got the authentication failed problem.
I am trying to ask help for checking the different parts between the workable configuration and mine.
Here is my configuration of outline:
![image](https://user-images.githubusercontent.com/73973812/191453760-ca9f65c2-ab05-4170-a58c-8da50116d4b8.png)

Keycloak auth:
![image](https://user-images.githubusercontent.com/73973812/191454073-829ff593-ebed-4908-8a4e-b35dd196ceca.png)

**Reply by AndreaPersico** [reply]
Hi!
Do you have a specific error to share in the thread?

In my configuration, the Outline Client is configured using other options:

<img width="699" alt="image" src="https://user-images.githubusercontent.com/49158551/191477989-a5c40aa3-2177-4260-aca6-dea27b8c14c3.png">

Best Regards!

PS: I'm still having issues related to MinIO uploading and I'm using AWS S3

**Reply by Wlmrsi** [reply]
@AndreaPersico Thanks for the reply.
If I don't add "openid" in OIDC_Scpoe, I will get the error as below:
``` WARN  [org.keycloak.services] (executor-thread-32) KC-SERVICES0091: Request is missing scope 'openid' so it's not treated as OIDC, but just pure OAuth2 request.```

but even I set the config as yours, then still same error message as below:

```
{"error":"Failed to obtain access token","stack":"InternalOAuthError: Failed to obtain access token\n   
at OAuth2Strategy._createOAuthError (/opt/outline/node_modules/passport-oauth2/lib/strategy.js:423:17)\n    
at /opt/outline/node_modules/passport-oauth2/lib/strategy.js:177:45\n    
at /opt/outline/node_modules/oauth/lib/oauth2.js:191:18\n    
at ClientRequest.<anonymous> (/opt/outline/node_modules/oauth/lib/oauth2.js:162:5)\n    
at ClientRequest.emit (node:events:526:28)\n    
at ClientRequest.emit (node:domain:475:12)\n    
at TLSSocket.socketErrorListener (node:_http_client:442:9)\n    
at TLSSocket.emit (node:events:526:28)\n    
at TLSSocket.emit (node:domain:475:12)\n    
at emitErrorNT (node:internal/streams/destroy:157:8)\n    
at emitErrorCloseNT (node:internal/streams/destroy:122:3)\n    
at processTicksAndRejections (node:internal/process/task_queues:83:21)","level":"error","message":"Error during authentication"}
```


It is not clear to find the problem or root cause, that is why I would like to know your steps/configuration to compare the different.....

**Reply by AndreaPersico** [reply]
Hi,
I have some questions that may be stupid, but I don't really know your environment:
1) are you using keycloak in dev mode or in production mode? 
2) are you using HTTPS enabled or disabled? (both on Outline and in Keycloak)
3) are you using a reverse proxy and self signed certificates?

Best Regards,
Andrea
